File: ios/Synapse/Core/Sync/SyncEngine.swift
import Foundation
import Observation

/// Keeps the local store current, in both directions.
///
/// Content comes down; a student's own work goes up. Neither the UI nor any
/// screen calls the network — they read `LocalStore`, and this is the only
/// thing that talks to the API. That separation is what makes every screen
/// work offline without knowing it is offline.
@MainActor
@Observable
final class SyncEngine {

    /// The catalogue documents a student is allowed to read.
    ///
    /// The same list as `STUDENT_READABLE_STATE` in `server/src/index.js`.
    /// Asking for anything outside it earns a 403, correctly — these are the
    /// documents the learning product is made of, and everything else on the
    /// server is operational data.
    static let catalogueKeys = [
        "synapse-academic-universities-v1",
        "synapse-course-curricula-v1",
        "synapse-module-schedules-v1",
        "synapse-admin-content-ledger-v4",
        "synapse-concept-graph-v2",
        "synapse-relation-types-v1",
        "synapse-taxonomy-tree-v4",
        "synapse-medical-library-taxonomy-v1",
        "synapse-medical-glossary-v1",
        "synapse-medical-evidence-published-v1",
        "synapse-plans-v1",
        "synapse-notification-campaigns-v1",
        "synapse-vouchers-v1",
        "synapse-system-colors-v1",
    ]

    /// The ledger: the one document every study surface is built from.
    static let ledgerKey = "synapse-admin-content-ledger-v4"
    /// The published evidence store. What resolves here is what a gated article
    /// may show — see `ArticleProjection`.
    static let evidenceKey = "synapse-medical-evidence-published-v1"
    /// Published concepts, for the annotations a reader may reveal.
    static let conceptGraphKey = "synapse-concept-graph-v2"
    /// The medical library taxonomy — the systems, disciplines, skills and
    /// knowledge branches a student browses by.
    static let medicalTaxonomyKey = "synapse-medical-library-taxonomy-v1"
    /// Universities and their years, for resolving a student's cohort.
    static let universitiesKey = "synapse-academic-universities-v1"
    /// The published timetable, per university/year/course.
    static let moduleSchedulesKey = "synapse-module-schedules-v1"
    /// The bilingual AR⇄EN glossary behind the taxonomy surface.
    static let glossaryKey = "synapse-medical-glossary-v1"
    /// Per-system colours, admin-editable. Fetched rather than hardcoded, so a
    /// new system in the admin console does not need an app release.
    static let systemColorsKey = "synapse-system-colors-v1"

    enum Status: Equatable {
        case idle
        case syncing
        /// Finished, with however many documents actually changed.
        case done(changed: Int, at: Date)
        /// Could not finish. The cache is still readable — it is simply not fresh.
        case failed(String)
    }

    private(set) var status: Status = .idle
    private(set) var pendingUploads = 0

    private let api: SynapseAPI
    private let store: LocalStore
    private var inFlight = false

    init(api: SynapseAPI, store: LocalStore) {
        self.api = api
        self.store = store
    }

    // MARK: - Refresh

    /// Fetch whatever changed, then drain the outbox.
    ///
    /// Safe to call on every launch and every foreground: the manifest makes
    /// "nothing changed" cost one small request rather than a full download.
    func refresh() async {
        guard !inFlight else { return }
        inFlight = true
        status = .syncing
        defer { inFlight = false }

        do {
            let changed = try await pullCatalogues()
            await drainOutbox()
            status = .done(changed: changed, at: Date())
        } catch {
            // A failed refresh is not a failed app. Say so, keep serving the
            // cache, and try again next time.
            status = .failed(Self.describe(error))
            await drainOutbox()
        }
    }

    /// Returns how many documents were actually re-fetched.
    private func pullCatalogues() async throws -> Int {
        let serverVersions: [String: Date?]
        do {
            serverVersions = try await api.stateManifest()
        } catch APIError.notFound, APIError.forbidden {
            // An API deployed before the manifest endpoint. Fall back to the
            // only behaviour available to it — fetch every catalogue — rather
            // than refusing to sync at all. Every key then reads as "not
            // mentioned" below, which is exactly that fallback.
            //
            // Both statuses mean the same thing here, and 403 is the one that
            // actually happens: without the route, `/api/state/:key` matches
            // the path with `key = "manifest"`, which is not in the readable
            // set, so a student is refused rather than 404'd. Treating only
            // 404 as "old server" left the app unable to sync at all against a
            // server that had simply not been redeployed.
            //
            // Safe to treat as absent: a student is never entitled to read a
            // document called `manifest`, so this cannot mask a real refusal.
            serverVersions = [:]
        }

        let localVersions = try await store.catalogueVersions()

        var changed = 0
        for key in Self.catalogueKeys {
            // Three cases, and they are genuinely different.
            guard let serverEntry = serverVersions.index(forKey: key).map({ serverVersions[$0].value }) else {
                // The manifest does not mention this key at all — an older
                // server that predates the endpoint. Fetch, because we cannot
                // tell whether it changed, and reading stale content is worse
                // than one extra request.
                if try await fetchCatalogue(key: key, expecting: nil) { changed += 1 }
                continue
            }

            guard let server = serverEntry else {
                // The manifest lists the key with no timestamp: nothing has
                // ever been written to it. There is no document to fetch, and
                // asking anyway would mean a wasted request on every launch for
                // every catalogue an install does not use.
                continue
            }

            // Skip only when the local copy carries the same stamp the server
            // reports. Anything else is fetched — guessing wrong here means a
            // student silently reads last month's content.
            if let local = localVersions[key] ?? nil, local == server { continue }

            if try await fetchCatalogue(key: key, expecting: server) { changed += 1 }
        }
        return changed
    }

    /// - Returns: whether the document was stored. A key a student may not read
    ///   is skipped permanently rather than retried.
    private func fetchCatalogue(key: String, expecting updatedAt: Date?) async throws -> Bool {
        do {
            let remote = try await api.state(JSONValue.self, key: key)
            guard let value = remote.value else { return false }

            let document = try JSONEncoder().encode(value)
            try await store.saveCatalogue(key: key, document: document, updatedAt: remote.updatedAt ?? updatedAt)

            if key == Self.ledgerKey {
                try await shredLedger(document)
            }
            return true
        } catch APIError.forbidden {
            // Not ours to read. Correct, and permanent — do not retry.
            return false
        }
    }

    /// Turn the ledger document into rows the library can browse and search.
    private func shredLedger(_ document: Data) async throws {
        let json = try JSONSerialization.jsonObject(with: document)
        let (items, skipped) = LedgerDecoder.decode(json)

        if skipped > 0 {
            // Worth saying out loud: a record the app cannot read is content a
            // student cannot see, and nothing else would report it.
            print("[sync] skipped \(skipped) unreadable ledger record(s)")
        }

        var searchTexts: [String: String] = [:]
        for item in items { searchTexts[item.id] = item.searchText }
        try await store.replaceItems(items, searchTexts: searchTexts)
    }

    // MARK: - The student's own work

    /// Record a change to one of the student's documents.
    ///
    /// Writes to the cache first and uploads after, so the app stays usable
    /// with no connection and the student never waits on a round trip to see
    /// their own edit.
    func write<Value: Encodable>(key: String, value: Value) async {
        assert(StateOwnership.isUserOwned(key), "\(key) is not a student-owned key — it would be refused by the server")

        do {
            let payload = try JSONEncoder().encode(value)
            try await store.enqueue(key: key, payload: payload)
            await refreshPendingCount()
            await drainOutbox()
        } catch {
            print("[sync] could not queue \(key): \(error)")
        }
    }

    /// Send everything waiting.
    func drainOutbox() async {
        let pending: [(key: String, payload: Data, queuedAt: Date, attempts: Int)]
        do {
            pending = try await store.pendingWrites()
        } catch {
            return
        }

        for entry in pending {
            do {
                let value = try JSONDecoder().decode(JSONValue.self, from: entry.payload)
                try await api.putUserState(key: entry.key, value: value)
                try await store.clearPending(key: entry.key, uploadedAt: entry.queuedAt)
            } catch APIError.forbidden {
                // The server will never accept this. Retrying forever would
                // block everything queued behind it.
                print("[sync] \(entry.key) refused; abandoning")
                try? await store.abandonPending(key: entry.key)
            } catch APIError.unauthorized {
                // Signed out mid-drain. Keep the work and stop — the rest of
                // the queue would fail the same way.
                return
            } catch {
                try? await store.recordFailure(key: entry.key, error: Self.describe(error))
            }
        }
        await refreshPendingCount()
    }

    private func refreshPendingCount() async {
        pendingUploads = (try? await store.pendingCount()) ?? 0
    }

    // MARK: - Sign-out

    /// Drop everything cached for the previous student.
    func clearForSignOut() async {
        try? await store.clearAll()
        pendingUploads = 0
        status = .idle
    }

    private static func describe(_ error: Error) -> String {
        switch error {
        case APIError.unauthorized: "Signed out"
        case APIError.forbidden: "Not permitted"
        case APIError.transient(let status): "Network error\(status.map { " (\($0))" } ?? "")"
        case APIError.malformed(let detail): "Unexpected response: \(detail)"
        default: error.localizedDescription
        }
    }
}

File: android/app/src/main/java/com/synapse/android/core/sync/SyncEngine.kt
package com.synapse.android.core.sync

import com.synapse.android.core.CortexJson
import com.synapse.android.core.adaptive.AdaptiveConfig
import com.synapse.android.core.adaptive.AdaptiveEvidenceStore
import com.synapse.android.core.adaptive.BlueprintNode
import com.synapse.android.core.adaptive.CoverageDebt
import com.synapse.android.core.adaptive.ReadinessResult
import com.synapse.android.core.api.ApiError
import com.synapse.android.core.api.RemoteState
import com.synapse.android.core.api.SynapseApi
import com.synapse.android.core.cache.LocalStore
import com.synapse.android.core.cache.PendingDocument
import com.synapse.android.core.library.LibraryMarks
import com.synapse.android.core.model.LedgerDecoder
import com.synapse.android.core.practical.PRACTICAL_PROGRESS_KEY
import com.synapse.android.core.progress.AttemptStore
import com.synapse.android.core.qbank.LiveSession
import java.time.Duration
import java.time.Instant
import java.time.YearMonth
import java.time.ZoneId
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.JsonNull

/**
 * What the last (or current) [SyncEngine.refresh] did.
 *
 * `Failed` deliberately carries no partial [Done]-style count: a refresh that
 * throws partway through has already left whatever it managed to fetch in
 * [LocalStore] — that is real progress, not a rollback — but reporting it as
 * a count here would suggest the whole pass succeeded when it did not.
 */
sealed interface SyncStatus {
    data object Idle : SyncStatus
    data object Syncing : SyncStatus
    data class Done(val changed: Int, val at: Instant) : SyncStatus
    data class Failed(val reason: String) : SyncStatus
}

/**
 * The only component in this app that is allowed to call the network.
 *
 * Every screen, ViewModel and repository reads through [LocalStore] — see the
 * module doc on that class. This is the thing that keeps it current: on
 * [refresh] it asks the server what changed since last time, downloads only
 * that, and reconciles it against whatever the device already holds; on
 * [write] it records a student's own edit locally and gets it back out to the
 * server, in order, even across process death (the queue lives in Room, not
 * in memory).
 */
class SyncEngine(
    private val api: SynapseApi,
    private val store: LocalStore,
) {
    private val _status = MutableStateFlow<SyncStatus>(SyncStatus.Idle)
    val status: StateFlow<SyncStatus> = _status.asStateFlow()

    // Guards refresh() against overlap. A plain flag read-then-set on
    // _status would race for real here — SynapseApi's calls resume on
    // OkHttp's own dispatcher thread, not necessarily the caller's — so this
    // needs an actual mutual-exclusion primitive, not just a checked field.
    private val refreshMutex = Mutex()

    /**
     * Ask the server what changed, pull it down, and reconcile it against
     * what is stored locally. A no-op if a refresh is already in flight —
     * see [refreshMutex].
     *
     * Order: the shared catalogue (manifest-diffed) → the per-student
     * documents the qbank and practical surfaces read, plus the attempt
     * ledger's index and its [AttemptStore.HISTORY_MONTHS] most recent month
     * shards (precedence-checked, since these are the ones a student can also
     * have edited offline) → whatever is still waiting in the outbox.
     */
    suspend fun refresh() {
        if (!refreshMutex.tryLock()) return
        try {
            _status.value = SyncStatus.Syncing
            // Carry any pre-rebrand `synapse…` cache forward before the first
            // read touches the store. Idempotent and self-limiting: once the
            // old rows are renamed there is nothing left for it to match.
            store.renameLegacyDocumentKeys(LEGACY_KEY_RENAMES)
            var changed = 0
            changed += syncCatalogue()
            changed += syncUserState()
            drain()
            _status.value = SyncStatus.Done(changed, Instant.now())
        } catch (e: CancellationException) {
            // Nothing failed and nothing finished -- the caller's scope died
            // mid-pass, which is what happens every time the student
            // backgrounds the app while [refreshWhenStale] has one running.
            // Leaving [_status] on [SyncStatus.Syncing] would leave Account
            // showing a spinner that never resolves for the life of the
            // process, and would tell the next [refreshWhenStale] that a
            // pass is still in flight.
            _status.value = SyncStatus.Idle
            throw e
        } catch (e: ApiError) {
            _status.value = SyncStatus.Failed(e.message ?: "sync failed")
        } finally {
            refreshMutex.unlock()
        }
    }

    /**
     * [refresh], unless one finished less than [minAge] ago.
     *
     * The app used to refresh exactly once per process, from `RootScreen`'s
     * arrival at `SignedIn`. An Android process outlives a great many
     * sessions on the same account: a student who answers questions on the
     * web at a desk and then picks their phone back up was reading a cache
     * that had not been asked about since the app first opened, with no way
     * to ask for one short of killing the app. So the foreground now
     * refreshes, and this is the throttle that makes that affordable --
     * without it, every task-switch, every rotation and every trip back from
     * the camera is a full manifest diff.
     *
     * Only a *completed* pass counts. A [SyncStatus.Failed] carries no
     * timestamp and a cancelled one resets to [SyncStatus.Idle], so coming
     * back to the foreground after a failure tries again immediately, which
     * is the moment it is most likely to work.
     */
    suspend fun refreshWhenStale(minAge: Duration) {
        val lastCompleted = (_status.value as? SyncStatus.Done)?.at
        if (lastCompleted != null && Duration.between(lastCompleted, Instant.now()) < minAge) return
        refresh()
    }

    /**
     * Records a student's own edit and gets it moving toward the server.
     *
     * The local copy and the outbox entry that carries it to the server are
     * written in one [LocalStore.putDocumentAndEnqueue] transaction — a
     * student must never see an edit as saved locally with nothing queued to
     * ship it, which two separate writes here would risk on a process death
     * between them. [drain] then runs opportunistically: most writes leave
     * with the same call that made them, but a write made without a network
     * never blocks on one — it simply waits in the queue for the next
     * [refresh] or [write] to drain it.
     */
    suspend fun write(key: String, json: String) {
        val existing = store.document(key)
        store.putDocumentAndEnqueue(key, json, existing?.serverUpdatedAt, Instant.now())
        drain()
    }

    /**
     * The same, for an edit that is only meaningful as a whole.
     *
     * [write] guarantees that one document and its outbox entry land
     * together; this guarantees the same across several documents at once.
     * Banking an attempt is the case that needs it: the month shard and the
     * index that counts it are two documents but one fact, and half of it
     * committed is worse than none of it -- a shard with no index entry is an
     * answer no headline figure counts, and an index with no shard counts an
     * answer no screen can show. Both are silent, and neither is recoverable
     * from the outside.
     *
     * Callers writing a single document should keep using [write]; there is
     * nothing to gain here and a list to build.
     */
    suspend fun writeAll(documents: List<Pair<String, String>>) {
        val pending = documents.map { (key, json) ->
            PendingDocument(key, json, store.document(key)?.serverUpdatedAt)
        }
        store.putDocumentsAndEnqueue(pending, Instant.now())
        drain()
    }

    /**
     * Replays the outbox to the server, oldest first, deleting each entry by
     * its own id the moment the server accepts it — never a blanket
     * "clear everything read", which would drop a write enqueued mid-drain.
     *
     * A document the server refuses outright ([ApiError.Forbidden]) is
     * dropped rather than retried forever — retrying changes nothing about
     * why it was refused. An entry whose own JSON cannot even be parsed is
     * dropped the same way: it is not going to parse any better on the next
     * drain, and leaving it in place would wedge every entry queued behind
     * it forever. Anything else that can fail keeps the entry (and
     * everything queued after it, so replay order is never scrambled) for
     * the next drain.
     */
    suspend fun drain() {
        for (entry in store.outbox()) {
            try {
                val value = CortexJson.parseToJsonElement(entry.json)
                api.writeState(entry.key, value)
                store.clearOutbox(entry.id)
            } catch (e: CancellationException) {
                throw e
            } catch (e: SerializationException) {
                store.clearOutbox(entry.id)
            } catch (e: ApiError.Forbidden) {
                store.clearOutbox(entry.id)
            } catch (e: ApiError) {
                break
            }
        }
    }

    /**
     * Diffs [CATALOGUE_KEYS] against the manifest and fetches only what
     * moved. A 404 on the manifest itself is a rollout fallback, not an
     * error — see the class doc on [SynapseApi.manifest] — and degrades to
     * fetching every catalogue document directly.
     *
     * Returns how many documents were actually written locally.
     */
    private suspend fun syncCatalogue(): Int {
        val manifest = try {
            api.manifest()
        } catch (e: ApiError.NotFound) {
            null
        }

        var changed = 0
        for (key in CATALOGUE_KEYS) {
            val needsFetch = manifest == null || store.document(key)?.serverUpdatedAt != manifest[key]
            if (!needsFetch) continue
            applyCatalogueDocument(key, api.readState(key))
            changed++
        }
        return changed
    }

    private suspend fun applyCatalogueDocument(key: String, remote: RemoteState) {
        val text = (remote.value ?: JsonNull).toString()
        store.putDocument(key, text, remote.updatedAt)
        if (key == LEDGER_KEY) {
            store.replaceLedger(LedgerDecoder.decode(text).items)
        }
    }

    /**
     * Pulls the per-student documents the qbank and practical surfaces read,
     * plus the attempt ledger's index and its [AttemptStore.HISTORY_MONTHS]
     * most recent month shards.
     * Unlike the catalogue there is no manifest for these — each is a
     * request of its own — so [StatePrecedence] is what keeps an offline
     * edit from being clobbered by a server copy that has not seen it yet.
     *
     * Returns how many documents were actually written locally.
     */
    private suspend fun syncUserState(): Int {
        var changed = 0
        for (key in userStateKeys()) {
            val remote = api.readState(key)
            val local = store.document(key)
            if (StatePrecedence.localCopyWins(local?.savedAt, remote.updatedAt)) continue
            if (remote.value == null && local == null) continue
            store.putDocument(key, (remote.value ?: JsonNull).toString(), remote.updatedAt)
            changed++
        }
        return changed
    }

    /**
     * The user-state keys one refresh pulls: the fixed documents, the attempt
     * index, and the last [AttemptStore.HISTORY_MONTHS] month shards.
     *
     * The window has to match the one [com.synapse.android.core.progress.AttemptLedger]
     * reads back, which is why both take the number from the same constant.
     * Fetching a shorter window than the ledger reads is invisible until a
     * student opens a fresh install and finds months of their own work
     * missing.
     */
    private fun userStateKeys(): List<String> {
        val zone = ZoneId.systemDefault()
        val currentMonth = YearMonth.now(zone)
        val months = (0 until AttemptStore.HISTORY_MONTHS)
            .map { currentMonth.minusMonths(it.toLong()) }
            .map { "%04d-%02d".format(it.year, it.monthValue) }
            .distinct()
            .sorted()
        // Adaptive Study's evidence ledger is month-sharded exactly like the
        // attempt ledger (`src/data/adaptive/evidenceLedger.ts`), so it is pulled
        // the same way: its index plus the same recent-month window
        // [com.synapse.android.core.adaptive.AdaptiveEvidenceStore] reads back.
        // The keys come from that object so the pull list cannot drift from the read.
        return USER_STATE_KEYS +
            AttemptStore.INDEX_KEY + months.map(AttemptStore::monthKey) +
            AdaptiveEvidenceStore.INDEX_KEY + months.map(AdaptiveEvidenceStore::monthKey)
    }

    companion object {
        /**
         * The shared catalogue: every key `STUDENT_READABLE_STATE` lists on
         * the server (`server/src/index.js`), in the same order. Enumerated,
         * not derived from a pattern — a key the server has not (yet)
         * granted students must not be guessed at here.
         */
        val CATALOGUE_KEYS: List<String> = listOf(
            "synapse-academic-universities-v1",
            "synapse-course-curricula-v1",
            "synapse-module-schedules-v1",
            "synapse-admin-content-ledger-v4",
            "synapse-concept-graph-v2",
            "synapse-relation-types-v1",
            "synapse-taxonomy-tree-v4",
            "synapse-medical-library-taxonomy-v1",
            "synapse-medical-glossary-v1",
            "synapse-medical-evidence-published-v1",
            "synapse-plans-v1",
            "synapse-notification-campaigns-v1",
            "synapse-vouchers-v1",
            "synapse-system-colors-v1",
            // Adaptive Study's admin-authored, student-read config and blueprint
            // (server `STUDENT_READABLE_STATE`). Referenced by named constant so
            // the read path and this pull list cannot spell the key differently.
            AdaptiveConfig.KEY, // nishany-adaptive-config-v1
            BlueprintNode.KEY, // nishany-adaptive-blueprints-v1
        )

        /** The content ledger's key — the one catalogue document that also gets shredded into rows [LocalStore] can query. */
        private const val LEDGER_KEY = "synapse-admin-content-ledger-v4"

        /**
         * The per-student documents the qbank, practical, notebook and
         * calendar surfaces read, plus the article-marks document the Reader
         * writes ([LibraryMarks.storageKey]). Enumerated for the same reason
         * [CATALOGUE_KEYS] is. A key outside this list (mastery, adaptive,
         * whiteboard, bookmarks, annotations) is deliberately out of scope
         * for this milestone.
         *
         * All entries are spelled `nishany…`, matching the web app post-rebrand.
         * The three that have a named constant (the write path uses it) are
         * referenced rather than re-spelled, so the pull list cannot drift from
         * what the feature code writes. Installs from before the rename still
         * hold these documents under their old `synapse…` keys in [LocalStore];
         * [LEGACY_KEY_RENAMES] carries that cached work forward on upgrade.
         */
        private val USER_STATE_KEYS: List<String> = listOf(
            LiveSession.KEY, // nishany.qbank.activeSession.v1 — src/pages/student/QuestionBank.tsx:232
            "nishany.qbank.marked.v1", // src/pages/student/QuestionBank.tsx:100
            "nishany.qbank.questionNotes.v1", // src/components/qbank/StudyRail.tsx:19
            LiveSession.SESSION_NAMES_KEY, // nishany.qbank.sessionNames.v1 — src/pages/student/QuestionBank.tsx:233
            PRACTICAL_PROGRESS_KEY, // nishany.practical.progress.v1 — src/data/practicalProgress.ts:17
            "nishany.notebook.notes", // src/pages/student/Notebook.tsx:62
            "nishany.calendar.tasks.v1", // src/data/tasks.ts:13
            // Adaptive Study's per-student documents. The month-sharded evidence
            // ledger (index + shards) is added in [userStateKeys]; these two are
            // whole documents. Named constants, referenced by the ViewModel read
            // path too, so the pull can never drift from the read. `boosts` and
            // the stored `plan` are deliberately out of scope: the dashboard reads
            // neither (no on-device session runner).
            CoverageDebt.KEY, // nishany.progress.adaptive.coverageDebt.v1
            ReadinessResult.KEY, // nishany.progress.adaptive.readiness.v1
            LibraryMarks.storageKey, // nishany.library.marks.v1 -- article highlights and sticky notes (M6)
        )

        /**
         * The rebrand renamed these five synced documents from `synapse…` to
         * `nishany…` (see [USER_STATE_KEYS]). An install from before the rename
         * still holds each one in [LocalStore] under its old key; this maps old
         * → new so a one-time on-device rename can carry that cached work
         * forward on upgrade, instead of stranding it under a key nothing reads.
         */
        private val LEGACY_KEY_RENAMES: List<Pair<String, String>> = listOf(
            "synapse.qbank.activeSession.v1" to "nishany.qbank.activeSession.v1",
            "synapse.qbank.marked.v1" to "nishany.qbank.marked.v1",
            "synapse.qbank.questionNotes.v1" to "nishany.qbank.questionNotes.v1",
            "synapse.qbank.sessionNames.v1" to "nishany.qbank.sessionNames.v1",
            "synapse.practical.progress.v1" to "nishany.practical.progress.v1",
        )
    }
}
